python/pipeline/pupil.py

```diff
@@ -1151,6 +1153,352 @@ def plot_fitting(key, start, end=-1, fit_type='Circle', fig=None, ax=None, mask_
+            raise PipelineException(msg)
+            
+        self.insert1({**key, 'conversion_factor': conversion_factor})
+
+
+@schema
+class PupilPeriodMethod(dj.Lookup):
+    definition = """ # Method to extract periods of dilation/constriction in pupil trace
+    period_method_id              : tinyint unsigned    # index of pupil period method
+    ---
+    period_method_description     : varchar(256)
+    """
+    
+    contents = [[1, 'Peak detection followed by filtering out events < 1sec or with an average speed < 0.02mm/sec. Any event comprised of over 15% NaNs is automatically dropped.'],
+                ]
+    
+
+@schema
+class PupilPeriods(dj.Computed):
+    definition = """ # Scan level statistics on dilations & constrictions
+    -> ProcessedPupil
+    -> PupilPeriodMethod
+    -> PupilUnitConversion
+    ---
+    dilation_freq                 : float               # dilations per second (within valid start/end)
+    constriction_freq             : float               # constrictions per second (within valid start/end)
+    num_pupil_periods             : smallint unsigned   # total number of dilation and constriction periods
+    """
+
+    class DilationConstriction(dj.Part):
+        definition = """ # Single period of dilation or constriction
+        -> PupilPeriods
+        period_idx                : smallint unsigned
+        ---
+        period_onset              : float               # period start time (in seconds)
+        period_offset             : float               # period stop time (in seconds)
+        peak_change_time          : float               # time (in seconds) at which peak change rate occurs
+        period_delta              : float               # overall change in size (in pixels, as a placeholder for now)
+        period_duration           : float               # period duration (in seconds)
+        """
+        
+    def _make_tuples(self, key):
+        
+        print(f'Populating PupilPeriods for {key}')
+
+        if key['period_method_id'] == 1:
+
+            ## DEFINE CONSTANTS
+            min_period_duration = 1      # sec
+            min_dilation_vel = 0.01      # mm/sec
+            min_constriction_vel = 0.01  # mm/sec (Note: This will be made negative in the code below)
+            max_nan_gap_percent = 0.15   # percentage of event made up by NaNs must be below this value
+
+            ## Fetch conversion factor
+            px_to_mm = (PupilUnitConversion & key).fetch1('conversion_factor')
+
+            ## Get timing data
+            pupil_times = (Eye & key).fetch1('eye_time')
+            pupil_fps = 1/np.nanmedian(np.diff(pupil_times))
+
+            ## Calculate peaks and valleys over nan-interpolated radii
+            ## np.diff(np.sign(derivative)) creates an array that's -1 when at a peak and 1 at valleys
+            ## (Apologies in advance for that opaque one-liner)
+            pupil_radii = (ProcessedPupil & key).fetch1('filtered_pupil_radius')
```

Comment:
Since this is used below in plot_pupil_periods, it might be a good idea to have this as its own helper function.